### PR TITLE
chore: update Python runtime and AWS CDK

### PR DESCRIPTION
**Pull request by ceholden**
## What I am changing

This PR addresses issue #37 by updating,

* Updates the (now deprecated) Python 3.8 runtime to Python ~3.12~ 3.11
    * I wanted to update to Python 3.12, but alas there is a CDK bug with Python 3.12 runtimes and `pipenv`, https://github.com/aws/aws-cdk/issues/30170. The recommendation is to downgrade to Python 3.11. Once this is fixed we should have a clear upgrade path to 3.12
* Updates the (now deprecated) AWS CDK v1 to CDK v2. This is required to support Python versions >=3.10
    * NB - This requires that our AWS accounts be bootstrapped for v2 of CDK, which should already be done because [hls-orchestration is using v2 of CDK](https://github.com/NASA-IMPACT/hls-orchestration/blob/843a044db947fefd2645e20da64eaef58bbe1c97/setup.py#L7)

I'm also hoping that some of the version bumps will prune down on the 82 Dependabot security warnings.

### Bumping Python versions

I bumped to the current latest Python runtime (3.12), but the CDK has a bug with that version so we're stuck on 3.11 for now. This required updating a few package dependencies. I saw that most of the `Pipfile`s used hard pinned versions for dependencies even though the versions are pinned in the lockfile, so I tried to keep to that convention. It would be easy to update this if we want to do that.

* Updating `psycopg2` to `>=2.9.9` to support Python 3.12
    * See [CHANGELOG](https://github.com/psycopg/psycopg2/blob/a805acf59f402f554e95624b5e27518169ca7715/NEWS#L14-L18) entry
* Remove now deprecated 3rd party `psycopg2-layer` from AWS Lambda layers. The `db` layer installs the same `psycopg2-binary` as the `psycopg2-layer` so we should be able to get that dependency from our own `db` layer
    * This only exists up until Python 3.9 and is now no longer maintained, https://github.com/jetbridge/psycopg2-lambda-layer?tab=readme-ov-file#no-longer-maintained
* Update `boto3` to `>=1.29.0` to support Python 3.12 for Boto3 and the Python 3.12 Lambda runtime
    * See [CHANGELOG](https://github.com/boto/boto3/blob/15c3dbc8758d64fec90664b8335111f410acfaee/CHANGELOG.rst#L2501)
* One of the packages didn't pin `moto` and I hit a breaking change in v5, so I went ahead and updated all of the packages that use Moto to the latest v5
    * From v4 to v5 they had a breaking change that migrated all `mock_[service]` decorator functions to just be `mock_aws`.
    * There was one other hangup - one of the `link_fetcher` tests failed because the mocked SQS queues apparently share the same state. I resolved this by naming each queue based on the test name.
    * See [CHANGELOG](https://github.com/getmoto/moto/blob/fa4352a9aab6e6e518f22975a8cccce1fff4a0e2/CHANGELOG.md#500) for v5
* Switched from `black/isort/flake8` to `ruff` as one linter/formatter to rule them all that is super fast

### Bumping AWS CDK

See migration guide for reference,
https://docs.aws.amazon.com/cdk/v2/guide/migrating-v2.html

The changes and related changes include,

* AWS consolidated all the non-beta functionality into just one package, simplifying our management of CDK dependencies
* Update imports from previously named `aws_cdk.core` (now in `aws_cdk` module)
* Remove CDK flags that have been removed in V2. These flags in V2 now behave as if the variable were set to "true" in v1
    * `core:enableStackNameDuplicates`
    * `aws-secretsmanager:parseOwnedSecretName`

## How you can test it

Run the very nice test suite!

Running from the repo root,
```bash
# make sure to have latest dependencies installed
$ make install
# with dummy postgres variables
$ PG_DB=foo PG_PASSWORD=foo PG_USER=foo make unit-tests
```

We can also help test that the CDK update is working without deploying by checking the CDK diff,
```
npx cdk diff
```

### CODE REVIEW

**Comment by ciaransweet** [comment]
@ceholden gimme a ping when this is ready for review! 

Always happy this little project is still alive and kicking 😍

**Comment by ceholden** [comment]
> @ceholden gimme a ping when this is ready for review!
> 
> Always happy this little project is still alive and kicking 😍

Thanks @ciaransweet! Appreciate the good work you did especially with docs, tests, and dev tooling. This sort of maintenance work feels a lot safer with all of that in place 💯 🚀 

I'm hoping to wrap this up ahead of adding the code I've written to feed the granule downloader from [ESA's notification system](https://documentation.dataspace.copernicus.eu/APIs/Subscriptions.html). From preliminary testing, moving to an event driven "link fetcher" should help us reduce download latency to ~3 minutes from new data publication time and also allow us to catch old granules that have been reprocessed by ESA. Not sure if you can access but I have a writeup of the experiment findings and plan for next steps in this ticket, https://github.com/NASA-IMPACT/hls_development/issues/300

I think I'll be fighting with some of the package version bumps some more, but would love your help reviewing when it's in a good place 😸

File: setup.py [hunk 1]
```diff
@@ -1,28 +1,10 @@
...
-    "aws-s3",
-    "aws-lambda-event-sources",
-    "aws-logs",
-]
-
+aws_cdk_version = "2.162.1"
 inst_reqs = [
-    *[f"aws_cdk.{x}=={aws_cdk_version}" for x in aws_cdk_reqs],
+    f"aws-cdk-lib=={aws_cdk_version}",
+    f"aws-cdk.aws-lambda-python-alpha=={aws_cdk_version}a0",
```
Comment (ceholden) [inline setup.py:7]: CDK v2 only has 1 main package but splits anything in "alpha" or "beta" off into separate packages. We're using `PythonFunction` from `aws-cdk.aws-lambda-python-alpha` so we need this additional package

**Comment by sharkinsspatial** [comment]
@ceholden Thanks so much for paying down all the technical debt we've accrued on this repo.  As you can see we don't need to make a lot of operational changes to the codebase since all the great work that @ciaransweet and @chuckwondo have put in just continues to crank away silently 💪

Since this is the second or third time we've needed to do major updates due to CDK version deprecation prior to making some functional changes, I'm wondering if we should start some type of regular CDK upgrade policy every 6 months across all repos 😆.

This looks great, I think the only thing left is update the code in the [app_integration.py](https://github.com/NASA-IMPACT/hls-sentinel2-downloader-serverless/blob/main/cdk/app_integration.py) which still has legacy CDK that throw an error when the CI attempts to deploy https://github.com/NASA-IMPACT/hls-sentinel2-downloader-serverless/blob/main/cdk/app_integration.py.

**Comment by ceholden** [comment]
Ayy I thought we were in the clear without needing to `cdk bootstrap` because other repositories use `cdk>=2` but alas we're not,

>This CDK deployment requires bootstrap stack version '6', found '0'. Please run 'cdk bootstrap'.

I have MCP access now, but I suspect the role I'll have won't permit this... I'll look into this and reach out with my findings.

**Comment by ceholden** [comment]
Thanks @chuckwondo for your help with updating our AWS credential configuration to use OIDC!

We are temporarily blocked on having this repository missing from the trust policy in the OIDC role. @chuckwondo submitted a service desk request to have the trust relationship updated as I will need a service desk request to submit service desk requests 😅

**Comment by ciaransweet** [comment]
> I will need a service desk request to submit service desk requests 😅

Spidey pointing at spidey

File: README.md [hunk 2]
```diff
@@ -21,9 +21,9 @@ This project aims to provide a serverless implementation of the current [HLS S2
 
 To develop on this project, you should install:
 
-* NVM [Node Version Manager](https://github.com/nvm-sh/nvm) / Node 12
+* NVM [Node Version Manager](https://github.com/nvm-sh/nvm) / Node 18
```
Comment (ciaransweet) [inline README.md:24]: Weeeew 6 node versions!

File: README.md [hunk 3]
```diff
@@ -105,7 +105,7 @@ A `Makefile` is available in the root of the repository to abstract away commonl
 
 **`make format`**
 
-> This will peform a run of `isort` and `black`, this **will** modify files if issues were found
+> This will perform a run of `isort` and `black`, this **will** modify files if issues were found
```
Comment (ciaransweet) [inline README.md:108]:
```suggestion
> This will perform a run of `ruff`, this **will** modify files if issues were found
```

Comment (ciaransweet) [reply README.md:108]: (Same applies to the lint one above this too)

**Review by ciaransweet — CHANGES_REQUESTED** [review]
Awesome job here @ceholden just a minor comment but looks gucci to me otherwise!

File: README.md (hunk 3 above)
Comment (ceholden) [reply README.md:108]: Nice catch, thanks!!

**Comment by ceholden** [comment]
Thanks to the help of @chuckwondo we're now past the hurdle of setting up CDK v2 bootstrap in `us-west-2` for the staging account 🎉. The deploy user we're using for dev requires some additional IAM permission grants that I'm working through before we can deploy the integration test stack. I _think_ I've gotten past that hurdle just now.

It looks like we also need to add a permissions boundary to the stacks because our deploy user's permissions to do things (e.g., create roles) depends on the permissions boundary being set for those deployed resources. Chuck's most recent PR has a nice example of what we need to do for this

**Review by ciaransweet — APPROVED** [review]

**Comment by chuckwondo** [comment]
> @chuckwondo thanks for your review and help on Friday! I did get the integration tests to run and succeed, but had a lot of trouble with misconfiguration in the staging account permissions boundaries. I'll ask around on Slack about next steps to fix this issue, but no code change should be required on this end

Can you elaborate on the misconfiguration? I'm not following, especially since int. tests have succeeded.

**Comment by ceholden** [comment]
@chuckwondo I'll followup on Slack with more info, but FYI I re-ran the integration tests to reproduce the issue after updating the Github environment. I'm going to edit the "dev" Github environment to remove the permissions boundary and re-run to ensure the integration tests are still passing

**Pull request merged**